### ISSUE TITLE
catalog zone issues

### ISSUE THREAD

**Issue by Habbie** [PowerDNS/pdns#11821]
# an edge case

fixed in #11825

1. ~~removing a zone (in my case, leaving the catalog zone with zero remaining members) does not bump the serial on the producer~~ this is not exactly true. More true: changes in member zone assignment to catalogs generally do not -reliably- bump the serial.
2. when the consumer eventually does process the new (empty) zone (so an update that removes the only present member zone) it crashes:
```
/usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/stl_iterator.h:994:16: runtime error: reference binding to null pointer of type 'const CatalogInfo'
    #0 0x559a1fe0270c in __gnu_cxx::__normal_iterator<CatalogInfo const*, std::vector<CatalogInfo, std::allocator<CatalogInfo> > >::operator*() const /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/stl_iterator.h:994:9
    #1 0x559a1fdc7d74 in catalogDiff(DomainInfo const&, std::vector<CatalogInfo, std::allocator<CatalogInfo> >&, std::vector<CatalogInfo, std::allocator<CatalogInfo> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /home/peter/projects/powerdns/pdns/pdns/slavecommunicator.cc:113:59
    #2 0x559a1fdb15b8 in catalogProcess(DomainInfo const&, std::vector<DNSResourceRecord, std::allocator<DNSResourceRecord> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) /home/peter/projects/powerdns/pdns/pdns/slavecommunicator.cc:419:10
    #3 0x559a1fd9f03f in CommunicatorClass::suck(DNSName const&, ComboAddress const&, bool) /home/peter/projects/powerdns/pdns/pdns/slavecommunicator.cc:791:12
    #4 0x559a1ee8857e in CommunicatorClass::retrievalLoopThread() /home/peter/projects/powerdns/pdns/pdns/communicator.cc:63:5
    #5 0x559a1ee8de67 in CommunicatorClass::go()::$_1::operator()() const /home/peter/projects/powerdns/pdns/pdns/communicator.cc:97:35
    #6 0x559a1ee8ddf1 in void std::__invoke_impl<void, CommunicatorClass::go()::$_1>(std::__invoke_other, CommunicatorClass::go()::$_1&&) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:60:14
    #7 0x559a1ee8dbf1 in std::__invoke_result<CommunicatorClass::go()::$_1>::type std::__invoke<CommunicatorClass::go()::$_1>(CommunicatorClass::go()::$_1&&) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:95:14
    #8 0x559a1ee8db44 in void std::thread::_Invoker<std::tuple<CommunicatorClass::go()::$_1> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/thread:264:13
    #9 0x559a1ee8da24 in std::thread::_Invoker<std::tuple<CommunicatorClass::go()::$_1> >::operator()() /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/thread:271:11
    #10 0x559a1ee8d7d8 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<CommunicatorClass::go()::$_1> > >::_M_run() /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/thread:215:13
    #11 0x7fe50854fecf  (/lib/x86_64-linux-gnu/libstdc++.so.6+0xceecf)
    #12 0x7fe508323ea6 in start_thread nptl/pthread_create.c:477:8
    #13 0x7fe50822cdee in clone misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:95

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/stl_iterator.h:994:16 in 
```

(note, for entry `#1` that' s line 113 in current master, my version has some extra debugging prints). Coverity also warned that this code might dereference something invalid:

![image](https://user-images.githubusercontent.com/525838/182591081-1bbc814b-cb85-4acc-bb98-4b8e46f10ab4.png)

This patch from @mind04 fixes that specific crash:
```diff
diff --git a/pdns/slavecommunicator.cc b/pdns/slavecommunicator.cc
index 6c573694a..06014a042 100644
--- a/pdns/slavecommunicator.cc
+++ b/pdns/slavecommunicator.cc
@@ -108,12 +108,12 @@ static bool catalogDiff(const DomainInfo& di, vector<CatalogInfo>& fromXFR, vect
       bool create{false};
       bool remove{false};
 
-      if ((xfr != fromXFR.end() && db == fromDB.end()) || *xfr < *db) { // create
+      if (xfr != fromXFR.end() && (db == fromDB.end() || *xfr < *db)) { // create
         ciCreate = *xfr;
         create = true;
         ++xfr;
       }
-      else if ((db != fromDB.end() && xfr == fromXFR.end()) || *db < *xfr) { // remove
+      else if (db != fromDB.end() && (xfr == fromXFR.end() || *db < *xfr)) { // remove
         ciRemove = *db;
         remove = true;
         ++db;
```

**Comment by Habbie** [comment]
Other catalog zone nits, collected here for now:
* producer/primary sometimes logs `Aug 03 12:07:57 Could not parse domain kind 'CONSUMER' as one of 'MASTER', 'SLAVE' or 'NATIVE'. Setting zone kind to 'NATIVE'`

**Comment by Habbie** [comment]
# broken SQL statements not noticed in testing

[this commit](https://github.com/PowerDNS/pdns/pull/11822/commits/d4e25b7601736d7cda1a8d5b19cf38b731532ef3) in #11822 fixes a typo that prevented catalog zones from working in gsqlite3 at all. My impression from reading the gsql-master/slave changes in regression-tests was that we would catch this, but we didn't.

**Comment by Habbie** [comment]
# pdns_control retrieve

After a bit of testing, I get the impression `pdns_control retrieve catalog.invalid` on a consumer does not cause catz processing.

**Comment by zeha** [comment]
# pgsql schema

I have the strong feeling that the new fields in the postgres schema should
a) just be TEXT, and not varchar with a length check. In PostgreSQL, there is no implementation difference. So if there are no powerdns reasons for this -> text.
b) all be NOT NULL. Mostly because if NULL is allowed, meaning has to be assigned to NULL vs. an empty string. To not break existing code (3rdparty or maybe builtin) then the defaults should become '', too.

Generally I feel that reordering columns is not worth the pain.

Transaction usage in the pgsql migration file could be optimised for production setups that do not want to stop pdns when updating.

**Comment by Habbie** [comment]
`''` in `domains.catalog` currently turns into a `DNSName(".")` object. If we can fix that (within this context), then that column can become NON NULL. I'll have to think about the others.

**Comment by zeha** [comment]
> `''` in `domains.catalog` currently turns into a `DNSName(".")` object. If we can fix that (within this context), then that column can become NON NULL. I'll have to think about the others.

Maybe it's valid for catalog to be NULL then, plus a check constraint ` <> ''` (i.e. prevent `.`)?

**Comment by Habbie** [comment]
# broken link

the link to the draft under https://doc.powerdns.com/authoritative/catalog.html#setting-up-catalog-zones has a spurious `)`

**Comment by Habbie** [comment]
# broken layout

https://docs.powerdns.com/authoritative/backends/index.html is broken now

**Comment by Habbie** [comment]
tl;dr: you make excellent points, but I've decided not to tackle them now.

> I have the strong feeling that the new fields in the postgres schema should
> a) just be TEXT, and not varchar with a length check. In PostgreSQL, there is no implementation difference. So if there are no powerdns reasons for this -> text. 

Let's do that broadly, separately.

> b) all be NOT NULL. Mostly because if NULL is allowed, meaning has to be assigned to NULL vs. an empty string. To not break existing code (3rdparty or maybe builtin) then the defaults should become '', too.

As discussed above, this would require some code changes. I'd very much love to do a round of schema cleanup after 4.7.

> Generally I feel that reordering columns is not worth the pain.

I agree, let's not do that any more. With the effort now spent testing on these migrations, I don't feel like changing them though.

> Transaction usage in the pgsql migration file could be optimised for production setups that do not want to stop pdns when updating.

Can you suggest/PR how?

> (If going for TEXT, maybe change all existing fields.)

I agree.

**Comment by Habbie** [comment]
Except for a potential improvement in the pgsql migration, all of this is done.

**Comment by Habbie** [comment]
> I agree, let's not do that any more. With the effort now spent testing on these migrations, I don't feel like changing them though.

I changed my mind. Simplifying the migrations is in #11836
